src/ra_log_wal.erl

```diff
@@ -714,46 +740,80 @@ dump_records(<<_:1/unsigned, 1:1/unsigned, _:22/unsigned,
-              ok
-    end.
+recover_wal_chunks(Fd, RecoveryChunkSize, Cache) ->
+    Chunk = read_from_wal_file(Fd, RecoveryChunkSize),
+    recover_records(Fd, Chunk, Cache, RecoveryChunkSize).
 
-recover_records(<<Trunc:1/unsigned, 0:1/unsigned, IdRef:22/unsigned,
+% Reached zeros indicating end of wal file
+recover_records(_, <<_:1/unsigned, 0:1/unsigned, _:22/unsigned, 
+                    IdDataLen:16/unsigned, _:IdDataLen/binary,
+                    _:32/integer, 0:32/unsigned, _/binary>>,
+                  _, _) ->
+    ok;
+% First record or different UID to last record
+recover_records(Fd, <<Trunc:1/unsigned, 0:1/unsigned, IdRef:22/unsigned,
                   IdDataLen:16/unsigned, UId:IdDataLen/binary,
                   Checksum:32/integer,
                   EntryDataLen:32/unsigned,
                   Idx:64/unsigned, Term:64/unsigned,
                   EntryData:EntryDataLen/binary,
-                  Rest/binary>>, Cache) ->
-    % first writer appearance in WAL
+                  Rest/binary>>,
+                Cache, RecoveryChunkSize) ->
     true = validate_and_update(UId, Checksum, Idx, Term, EntryData, Trunc),
-    % TODO: recover writers info, i.e. last index seen
-    recover_records(Rest,
-                    Cache#{IdRef =>
-                           {UId, <<1:1/unsigned, IdRef:22/unsigned>>}});
-recover_records(<<Trunc:1/unsigned, 1:1/unsigned, IdRef:22/unsigned,
+    recover_records(Fd, Rest,
+                    Cache#{
+                        IdRef => {UId, <<1:1/unsigned, IdRef:22/unsigned>>}
+                    }, RecoveryChunkSize);
+
+ % TODO: recover writers info, i.e. last index seen
+% Same UID as last record
+recover_records(Fd, <<Trunc:1/unsigned, 1:1/unsigned, IdRef:22/unsigned,
                   Checksum:32/integer,
                   EntryDataLen:32/unsigned,
                   Idx:64/unsigned, Term:64/unsigned,
                   EntryData:EntryDataLen/binary,
-                  Rest/binary>>, Cache) ->
+                  Rest/binary>>,
+                Cache, RecoveryChunkSize) ->
     #{IdRef := {UId, _}} = Cache,
     true = validate_and_update(UId, Checksum, Idx, Term, EntryData, Trunc),
+    recover_records(Fd, Rest, Cache, RecoveryChunkSize);
+% Reached end of chunk, with no remainder, time to read the next chunk
+recover_records(Fd, <<>>, Cache, RecoveryChunkSize) ->
+    Chunk = read_from_wal_file(Fd, RecoveryChunkSize),
+    case Chunk of
+        <<>> -> ok;
+        _ -> recover_records(Fd, Chunk, Cache, RecoveryChunkSize)
+    end;
+% Reached the of chunk, with some remainder, time to read the next chunk
+recover_records(Fd, Chunk, Cache, RecoveryChunkSize) ->
+    NextChunk = read_from_wal_file(Fd, RecoveryChunkSize),
+    
+    case NextChunk of 
+        <<>> ->
+            % we were expecting more to be read and there wasn't meaning
+            % we have a partial record. This should never happen.
+            ?WARN("Partial record lost", []),
```

Comment:
This isn't necessarily the case as you'd arrive here if you had a partial zero-padded remainder which isn't a lost record

Reply:
Maybe lower to a debug level and prepend the message with "wal: " or something to make it clear where it is coming from.

Reply:
As far as I see, this should only match the situation where we have a partial record, and we're at the end of the file. A remainder cannot be zero padded (in the sense that the zeros indicate that we've reached the end of the data). A remainder can only exist if the chunk did not have the whole record, if it has zeros at the end then we have a bad data situation.

Reply:
Surely if you write records into a zero padded wal file and you happen to have a couple of empty bytes left at the end you'd reach this clause without it being an error?

Reply:
Ok, i've made some tweaks, I don't think it should be a warn as it may just worry users when it isn't a big deal, any reminder at the end of the file that cannot be parsed can safely be discarded